nncf/common/graph/patterns.py

```diff
@@ -0,0 +1,167 @@
+                # A: (a) (b)
+                # B: (c) (d)
+                #              (a)  (a_copy)  (b)    (b_copy)
+                # A + B ---->   |       |      |        |
+                #              (c)     (d)  (c_copy) (d_copy)
+                #
+                subgraph_copy = create_copy_of_subgraph(self_subgraph)
+                other_subgraph_copy = create_copy_of_subgraph(other_subgraph)
+                subgraph_copy = _add_second_subgraph_to_first_with_connected_edge(subgraph_copy, other_subgraph_copy)
+                add_subgraph_to_graph(final_graph, subgraph_copy)
+
+        final_pattern = copy.deepcopy(self)
+        final_pattern._graph = final_graph
+        return final_pattern
+
+    def __or__(self, other):
+        """
+        Add other's DiGraph nodes to self's DiGraph.
+        Other's DiGraph will be placed as a new weakly connected components
+        """
+        new_pattern = copy.deepcopy(self)
+        other_copy = create_copy_of_subgraph(other.graph)
+        add_subgraph_to_graph(new_pattern._graph, other_copy)
+        return new_pattern
+
+    def __eq__(self, other):
+        return ism.is_isomorphic(self.graph, other.graph)
+
+    @property
+    def graph(self):
+        return self._graph
+
+    def add_pattern_alternative(self, other: 'GraphPattern') -> None:
+        other_graph = other.graph
+        self_graph = self.graph
+        for node in other_graph.nodes:
+            if node in self_graph.nodes:
+                assert False
+            self_graph.add_node(node, type=other_graph.nodes[node]['type'])
+        for edge in other_graph.edges:
+            self_graph.add_edge(edge[0], edge[1])
+
+    def join_patterns(self, other: 'GraphPattern', edges: Optional[List[Tuple[Hashable, Hashable]]] = None) -> None:
+        if edges is None:
+            first_node_other = list(nx.topological_sort(other.graph))[0]
+            assert other.graph.in_degree(first_node_other) == 0
+            last_node_self = list(nx.topological_sort(self.graph))[-1]
+            assert self.graph.out_degree(last_node_self) == 0
+            edges = [[last_node_self, first_node_other]]
+        for edge in edges:
+            assert edge[0] in self.graph.nodes
+            assert edge[1] in other.graph.nodes
+        add_subgraph_to_graph(self.graph, other.graph)
+        for edge in edges:
+            self.graph.add_edge(edge[0], edge[1])
+
+    def add_node(self, t: List[str]) -> int:
+        self.graph.add_node(GraphPattern.NODE_COUNTER, type=t)
+        GraphPattern.NODE_COUNTER += 1
+        return GraphPattern.NODE_COUNTER - 1
+
+    def add_edge(self, u_name, v_name) -> None:
+        self.graph.add_edge(u_name, v_name)
+
+    def get_weakly_connected_subgraphs(self) -> List[nx.DiGraph]:
+        return [self.graph.subgraph(c) for c in nx.weakly_connected_components(self.graph)]
+
+    def dump_graph(self, path: str) -> None:
+        nx.drawing.nx_pydot.write_dot(self.graph, path)
```

Comment:
Also add a function to better visualize the graph, so that the visualized graph has the attributes to be matched and their values visible.

Reply:
For better visualization, I added a field "label", which Graphviz uses as the name of the node. So, the dumped pattern graph now looks like this:
![image](https://user-images.githubusercontent.com/32935044/123222081-40caf780-d4d8-11eb-98cf-6def10220865.png)

Reply:
The exact values that are being matched are still not being visualized, add a sort of a legend for the capitalized identifiers.

Reply:
There is such an opportunity to support 2 options for dumping the graph. The first variant as is. The second variant, dump each pattern in a separate dot file with the relevant name of the file.

Reply:
I introduce a new class `QuantizationIgnorePatterns`, which stores all registered patterns. This class has 3 functions for visualizations:
1) Dump each pattern in a dir separately
2) Dump all matching patterns in a one file
3) Dump a specific pattern